add and use `has_xfree`

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
fixes https://github.com/sympy/sympy/issues/23775 by removing auto-expansion by linsolve (while still handling Eq)
alternative to #23748

#### Brief description of what is fixed or changed

`has_free` is slower than needed for some cases since it does sub-expression matches:
```python
>>> eq = f(x+y+1)
>>> eq.has_free(f)
True
>>> eq.has_free(x + 1)
True
```

`has_xfree` is more strict and will only return True for full expression matches (similar to the way that `xreplace` makes replacements), so both of the above return False with `has_xfree`.

#### Other comments

`linsolve` and related functions that extract coefficients from linear systems fail when any term or factor depends on a provided symbol. Rather than providing a limited workaround for cases like treating `f(x)` as independent of `f(x).diff()`, those needing to solve such systems can replace all symbol-dependent objects with dummy symbols. The `recast_to_symbols` can do this:

```python
>>> from sympy.solvers.solveset import recast_to_symbols
>>> eqs = Tuple(f(x)+f(x).diff())
>>> e, k, d = recast_to_symbols(eqs, eqs.atoms(Derivative))
>>> linsolve(e, [f(x)]).xreplace(d)
{(-Derivative(f(x), x),)}
```
Alternatively (though hackish) one could temporarily define Derivative's `has_xfree` to return None
```python
>>> Derivative.has_xfree = lambda a, *b: None
>>> linsolve(eqs, [f(x)])
{(-Derivative(f(x), x),)}
>>> del Derivative.has_xfree
```
I tried to define `Derivative.bound_symbols` as `set(Derivative.variables)` so `iterfreeargs` would not return the function itself. This worked wrt `linsolve`, but many other tests failed. I am not sure of the implications, but it *seems* like the variables of differentation should be considered bound.
```diff
diff --git a/sympy/core/function.py b/sympy/core/function.py
index 551cf41fe8..0f0c6295f4 100644
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -1480,6 +1480,10 @@ def __new__(cls, expr, *variables, **kwargs):
             expr = factor_terms(signsimp(expr))
         return expr

+    @property
+    def bound_symbols(cls):
+        return set(cls.variables)
+
     @property
     def canonical(cls):
         return cls.func(cls.expr,
```
then
```python
>>> from sympy.core.traversal import *
>>> list(iterfreeargs(f(x).diff(x,2)))
[Derivative(f(x), (x, 2)), 2]  --> in master: [Derivative(f(x), (x, 2)), f(x), (x, 2), x, x, 2]
>>> list(iterfreeargs(Integral(x,(x,1,2))))
[Integral(x, (x, 1, 2)), 1, 2]
>>> linsolve([f(x)+f(x).diff()], [f(x)])
{(-Derivative(f(x), x),)}
>>> integrate(Derivative(f(x), x), x)
x*Derivative(f(x), x)  --> in master: f(x)
>>> integrate(x**3*Derivative(f(x), (x, 4)))
x**4*Derivative(f(x), (x, 4))/4 --> in master: x**3*Derivative(f(x), (x, 3)) - 3*x**2*Derivative(f(x), (x, 2)) + 6*x*Derivative(f(x), x) - 6*f(x)
```

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
  * `linsolve` no longer fully expands expressions so any nonlinear terms that would cancel after expansion now raise
  * `linear_coeffs` is more efficient and can return results as a dictionary (which is more economical for sparse equations)
  * `linear_eq_to_matrix` can now handle system expressed in terms of functions
* core
  * `Basic.has_xfree` has been added. Like `xreplace`, it targets only full expression matching
<!-- END RELEASE NOTES -->

smichr committed:
add has_xfree

smichr committed:
more efficient coefficient extraction

smichr commented:
@oscarbenjamin , I don't know if you see the notifications for request. I have no idea why it marked me as making the request. I didn't (intentionally) do so. I wonder if this is ever the source of ping that were frowned upon in the discussion about the new template proposal for PRs. Anyway, removal of false request should not to be taken as positive or negative. I know you will take a look when you can.

smichr commented:
@sylee957 , I can't see the problem with the linsolve bullet list, can you?

oscarbenjamin commented on sympy/solvers/ode/tests/test_systems.py:
```diff
@@ -1071,8 +1071,8 @@ def test_sysode_linear_neq_order1_type2():
-    sol6 = [Eq(f(x), C1*exp(-4*x)*Rational(-4, 5) + C2*exp(-9*x)),
-            Eq(g(x), C1*exp(-4*x)),
+    sol6 = [Eq(f(x), C2*exp(-4*x)*Rational(-4, 5) + C1*exp(-9*x)),
+            Eq(g(x), C2*exp(-4*x)),
```
I can't work out why this has changed.

smichr replied:
Puzzled me, too. I think it has to do with the ordering of symbols that are used to replace the objects. `solve` takes care to make the "symbols" canonical before assigning dummies, but the `systems` code did not:
```python
rep = {func_deriv: Dummy() for func_deriv in funcs_deriv}  # ordered(funcs_deriv)
```
Just a hunch.

oscarbenjamin replied:
The code that does this calls `linear_eq_to_matrix` but if I read this correctly it should preserve the order of `eqs` and `syms` so the output from `linear_eq_to_matrix` should be unchanged:
https://github.com/sympy/sympy/blob/39de9a2698ad4bb90681c0fdb70b30a78233145f/sympy/solvers/ode/systems.py#L471-L484

smichr replied:
I now suspect that it has to do with not returning a 0 coefficient from `_lin_eq2dict`. Here are equations being renumbered during the test involving `sol6`:
master 
```python
[Eq(f(x), (-4*_Dummy_148/5 - 4*Integral(0, x)/5)*exp(-4*x) + (_Dummy_147 + Integral(0, x))*exp(-9*x)), Eq(g(x), (_Dummy_148 + Integral(0, x))*exp(-4*x)), Eq(h(x), (_Dummy_149 + Integral(1, x))*exp(x))]
[Eq(f(x), (-4*C1/5 - 4*Integral(0, x)/5)*exp(-4*x) + (C2 + Integral(0, x))*exp(-9*x)), Eq(g(x), (C1 + Integral(0, x))*exp(-4*x)), Eq(h(x), (C3 + Integral(1, x))*exp(x))]
```
PR
```python
[Eq(f(x), _Dummy_121*exp(-9*x) - 4*_Dummy_122*exp(-4*x)/5), Eq(g(x), _Dummy_122*exp(-4*x)), Eq(h(x), (_Dummy_123 + Integral(1, x))*exp(x))]
[Eq(f(x), C1*exp(-9*x) - 4*C2*exp(-4*x)/5), Eq(g(x), C2*exp(-4*x)), Eq(h(x), (C3 + Integral(1, x))*exp(x))]
```
That factor of `Integral(0, x)` might through things off.

smichr replied:
It changes form depending on whether Matrix or SparseMatrix is used to send back the results from `linear_eq_to_matrix`. Simply wrapping Matrix around `SparseMatrix(*shape, dat)` will cause the results to change. This is the minimal change necessary to toggle between the two results (but it is not the root cause).

smichr committed:
don't integrate 0 wrt t

It doesn't hurt, but it changes the form of the solution when a Sparse or Dense matrix is used because the Sparse automatically doesn't provide zeros (so they aren't integrated) while Dense does.

smichr commented on sympy/solvers/ode/systems.py:
```diff
@@ -464,22 +463,16 @@ def linear_ode_to_matrix(eqs, funcs, t, order):
         raise ODEOrderError(msg.format(order))
 
-    As = []
+    As = []#
```
```suggestion
    As = []
```

smichr commented on sympy/solvers/ode/systems.py:
```diff
@@ -464,22 +463,16 @@ def linear_ode_to_matrix(eqs, funcs, t, order):
         # eqs is minus the remainder of the equations.
         try:
-            Ai, b = linear_eq_to_matrix(eqs, syms)
+            Ai, b = linear_eq_to_matrix(eqs, syms)#
```
```suggestion
            Ai, b = linear_eq_to_matrix(eqs, syms)
```

smichr commented:
I need help resolving the SPHINX error.

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -2760,16 +2771,23 @@ def linsolve(system, *symbols):
-    * An error is raised if, after expansion, any nonlinearity
-      is detected:
+    * An error is raised if any nonlinearity is detected, even
+    if it could be removed with expansion
```
I'm guessing that the indentation here is the sphinx problem:
```
/home/runner/work/sympy/sympy/sympy/solvers/solveset.py:docstring of sympy.solvers.solveset.linsolve:169: WARNING: Bullet list ends without a blank line; unexpected unindent.
```

smichr replied:
I used the editor to search for leading space followed by `*`, looking for this case. Not sure how I missed it! Thanks.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -2760,16 +2771,23 @@ def linsolve(system, *symbols):
 
-    * An error is raised if, after expansion, any nonlinearity
-      is detected:
+    * An error is raised if any nonlinearity is detected, even
+    if it could be removed with expansion
```
```suggestion
    * An error is raised if any nonlinearity is detected, even
      if it could be removed with expansion
```

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1284,8 +1311,26 @@ def has_free(self, *patterns):
+            rv = self.has_xfree(s)
+            if rv:
+                return rv
+        except TypeError:
```
It's usually a bad idea to catch `TypeError`. Why is it being caught here?

smichr replied:
In case of an un unhashable pattern which is not handled until we get to `_has`.

oscarbenjamin replied:
What is an example of an unhashable pattern? Would they not be hashable if sympified?

oscarbenjamin replied:
I would still prefer not to catch `TypeError`. It is not a good class of exceptions to catch. At least if we are catching it then the `try` block needs to be restricted in scope to a single operation that is the one where `TypeError` is expected and there should be comments explaining precisely why `TypeError` is being caused.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1284,8 +1311,26 @@ def has_free(self, *patterns):
+        if not patterns:
+            return False
+        p0 = patterns[0]
+        if len(patterns) == 1 and iterable(p0) and not isinstance(p0, Basic):
```
Why would this not be Basic? Is it because it needs to be sympified?

smichr replied:
No, because the user may fogetting to unpack an iterable. Sometimes you do (here) and sometimes you don't (`has_xfree`). There are several functions like this besides these. We could just immediately return False (GIGO).
```
            # Basic can contain iterables (though not non-Basic, ideally)
            # but don't encourage mixed passing patterns
            raise ValueError(filldedent('''
                Expecting 1 or more Basic args, not a single
                non-Basic iterable. Don't forget to unpack
                iterables: `eq.has_free(*patterns)`'''))
```

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -2594,14 +2600,19 @@ def linear_eq_to_matrix(equations, *symbols):
+    rhs = [-i for i in c]
+    del c
+    A = SparseMatrix(*shape, dat)
+    b = SparseMatrix(n, 1, rhs)
```
I pretty much want to abolish SparseMatrix so it would be better not to add new uses of it.

oscarbenjamin commented on sympy/solvers/ode/tests/test_systems.py:
```diff
@@ -1071,8 +1071,8 @@ def test_sysode_linear_neq_order1_type2():
-    sol6 = [Eq(f(x), C1*exp(-4*x)*Rational(-4, 5) + C2*exp(-9*x)),
-            Eq(g(x), C1*exp(-4*x)),
+    sol6 = [Eq(f(x), C2*exp(-4*x)*Rational(-4, 5) + C1*exp(-9*x)),
+            Eq(g(x), C2*exp(-4*x)),
```
I'm still unclear why this has changed. I don't see a reason that the changes here should change in any way the output of `linsolve` or `linear_eq_to_matrix`.

smichr replied:
unresolved terms that are zero get numbered; they are no longer retained

smichr replied:
You can see an example of an expression with `Integrate(0, t)` in the changed tests.

oscarbenjamin replied:
Okay, I see.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1258,6 +1258,33 @@ def has(self, *patterns):
         """
         return self._has(iterargs, *patterns)
 
+    def has_xfree(self, s):
```
Maybe this should be type-hinted as `s: set[Basic]`.

smichr replied:
I don't follow you. `s` should be a set, not Basic.

oscarbenjamin replied:
It should be a `set` of `Basic` which is what `set[Basic]` means in typing syntax.

oscarbenjamin commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -194,16 +199,30 @@ def _lin_eq2dict(a, symset):
-    elif not a.has_free(*symset):
+    if a.is_Equality:
+        (coeff, terms), (cR, tR) = [_lin_eq2dict(ai, symset)
+            for ai in a.args]
```
Maybe equalities should be handled at a higher level before calling this function.

smichr replied:
I had already considered that. It is here because otherwise anyone using any of the linear coefficient extracting routines would have to duplicate this code to carefully make sure there is no linearity and then allow cancellation of linear terms.

oscarbenjamin replied:
It could be handled by `_linear_eq_to_dict` for example.

The checking for zeros probably needs to be handled more broadly than just for `Equality`.

smichr commented:
Instead of adding a new function we could add a new handling to `has_free` so that a single set would be treated with literal logic (like `has_xfree`). Any other iterable that is not Basic would raise. Any unpacked list would get the full treatment.

oscarbenjamin commented:
> Instead of adding a new function we could add a new handling to `has_free` so that a single set would be treated with literal logic (like `has_xfree`).

The problem is the ambiguity in the case of a `FiniteSet` since a `set` sympifies to a `FiniteSet`:
```python
In [2]: Complement({x, y}, {z}).has_free({x, y})
Out[2]: True
```

oscarbenjamin commented:
I just think that it's better not to have an ambiguous API like that. Too many functions and methods in SymPy accept overly complicated combinations of different possible types of inputs making it difficult to even define what should or should not be expected and making it very hard to change the implementation of how they decide what to do. All that is needed here is a function that takes a set of expressions.

smichr commented:
> I just think that it's better not to have an ambiguous API like that.

I'm ok with that, too. Is there anything else that needs to be changed here?

oscarbenjamin commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -126,53 +129,70 @@ def sympy_dict_to_dm(eqs_coeffs, eqs_rhs, syms):
+    symset = set(syms)
+    for i, e in enumerate(eqs):
+        if e.is_Equality:
+            (coeff, terms), (cR, tR) = [_lin_eq2dict(ai, symset)
+                for ai in e.args]
```
A loop to avoid repeating a line just twice is not really a good application of DRY. It's a lot easier to understand the straight-forward repetitive code:
```python
coeff, terms = _lin_eq2dict(e.lhs, symset)
cR, tR = _lin_eq2dict(e.rhs, symset)
```

smichr replied:
```suggestion
            coeff, terms = _lin_eq2dict(e.lhs, symset)
            cR, tR = _lin_eq2dict(e.rhs, symset)
```

oscarbenjamin commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -126,53 +129,70 @@ def sympy_dict_to_dm(eqs_coeffs, eqs_rhs, syms):
+                if k in terms:
+                    terms[k] -= v
+                else:
+                    terms[k] = v
```
Should this not be `-v`?

smichr replied:
Hmm.. that should be tested.

oscarbenjamin commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -126,53 +129,70 @@ def sympy_dict_to_dm(eqs_coeffs, eqs_rhs, syms):
     if a in symset:
         return S.Zero, {a: S.One}
-    elif a.is_Add:
+    if a.is_Add:
```
Why have you changed these `elif` to `if`? It's much clearer to use `elif` to show that all cases are mutually exclusive without needing to scan through to see that every control path either returns or raises.

smichr replied:
Preference. If every block clearly ends with a return (or starts with an elif) coverage is complete.

smichr commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -126,53 +129,70 @@ def sympy_dict_to_dm(eqs_coeffs, eqs_rhs, syms):
+                if k in terms:
+                    terms[k] -= v
+                else:
+                    terms[k] = v
```
```suggestion
                    terms[k] = -v
```

smichr commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -126,53 +129,70 @@ def sympy_dict_to_dm(eqs_coeffs, eqs_rhs, syms):
     if a in symset:
         return S.Zero, {a: S.One}
-    elif a.is_Add:
+    if a.is_Add:
```
```suggestion
    elif a.is_Add:
```

smichr commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -183,7 +203,7 @@ def _lin_eq2dict(a, symset):
         terms = {sym: Add(*coeffs) for sym, coeffs in terms_list.items()}
         return coeff, terms
-    elif a.is_Mul:
+    if a.is_Mul:
```
```suggestion
    elif a.is_Mul:
```

smichr commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -194,16 +214,16 @@ def _lin_eq2dict(a, symset):
-    elif a.is_Equality:
-        return _lin_eq2dict(a.lhs - a.rhs, symset)
-    elif not a.has_free(*symset):
+    if not a.has_xfree(symset):
```
```suggestion
    elif not a.has_xfree(symset):
```

smichr commented on sympy/polys/matrices/linsolve.py:
```diff
@@ -194,16 +214,16 @@ def _lin_eq2dict(a, symset):
         return a, {}
-    else:
-        raise PolyNonlinearError
+    raise PolyNonlinearError('nonlinear term: %s' % a)
```
```suggestion
    else:
        raise PolyNonlinearError('nonlinear term: %s' % a)
```

smichr commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1502,12 +1513,11 @@ def test_linsolve():
-    # and other nonlinearity does not raise an error
+    # linsolve does not allow expansion (real or implemented)
+    # to remove singularities, but it will cancel linear terms
     assert linsolve([Eq(x, x + y)], [x, y]) == {(x, 0)}
```
```suggestion
    assert linsolve([Eq(x, x + y)], [x, y]) == {(x, 0)}
    assert linsolve([Eq(x + x*y, 1 + y)], [x]) == {(1,)}
    assert linsolve([Eq(1 + y, x + x*y)], [x]) == {(1,)}
```

oscarbenjamin commented on sympy/core/tests/test_expr.py:
```diff
@@ -1624,12 +1625,31 @@ def test_has_free():
+    assert Piecewise((s, x > 3), (4, True)).has_free(s)
+    assert not Piecewise((1, x > 3), (4, True)).has_free(s)
+    # can't make set of these, but fallback will handle
+    assert not x.has_free(y, [])
```
Shouldn't this just give a TypeError?

smichr replied:
Being empty is tested before being a non-empty container.

smichr replied:
I'm not sure that it is yet a valid assumption that all Basic args are hashable. If it is, then an error can be raised.

oscarbenjamin replied:
Basic args should be Basic and therefore should be hashable. If they aren't then that's a bug and it should be fixed. We shouldn't try to work around bugs like that because it just makes them less visible. There shouldn't be any code that attempts to allow for Basic to have non-Basic args.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1258,6 +1258,33 @@ def has(self, *patterns):
+        """
+        # protect O(1) containment check by requiring:
+        if not type(s) in (dict, set):
+            raise ValueError('expecting set or dict argument')
```
I guess TypeError is more appropriate. The type hint in the signature suggests that only `set` is allowed. I suggest only allowing `set` here. I don't see a strong benefit in anything else.

smichr replied:
I put the dict there because you indicated that dict or set would have O(1) lookup. And there is no reason to force the user to convert dict to set if it can be used directly. I can make up a case where someone would want to pass a dict, but I'm not sure *I* would do so in practice. But I have appreciated the flexibility of Python to just "do the right thing" if no rules are broken...and to enforce as few rules as necessary.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1258,6 +1258,33 @@ def has(self, *patterns):
         """
         return self._has(iterargs, *patterns)
 
+    def has_xfree(self, s: set[Basic]):
```
If we're type-hinting then we can also add the return type `-> bool`.

smichr replied:
```suggestion
    def has_xfree(self, s: set[Basic]) -> bool:
```

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1284,8 +1311,27 @@ def has_free(self, *patterns):
+        else:
+            rv = self.has_xfree(s)
+            if rv:
+                return rv
```
I always find this sort of thing confusing. Why `return rv` and not just an explicit `return True` since we know that `rv` is a `bool`?

smichr replied:
Because then I don't have to think about if it's S.true or True. It's right.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1284,8 +1311,27 @@ def has_free(self, *patterns):
+        try:
+            s = set(patterns)
+        except TypeError:
+            pass  # patterns had a non-hashable element
```
As I've said before I would prefer not to catch `TypeError`. There are too many places that do this. A `TypeError` should almost always mean a bug and so the error should not generally be caught.

smichr replied:
Please make a suggestion since there is no other error to catch and the scope has been limited to a single obvious case of why it might be raised. 

The reason this is here is because if we can make a set of the items and this test passes then it is a quick exit. And if not (either unhashable items or pattern not found) then we go to the next step where unhashable things are thrown out. In case of unhashable I don't want to let this raise for the user because the next step will ignore them (or could its own TypeError).

oscarbenjamin replied:
Why we can't we just let this TypeError bubble up?

smichr replied:
I guess we can see what fails...will make the change

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1284,8 +1311,27 @@ def has_free(self, *patterns):
+        if len(patterns) == 1 and iterable(p0) and not isinstance(p0, Basic):
+            # Basic can contain iterables (though not non-Basic, ideally)
+            # but don't encourage mixed passing patterns
+            raise ValueError(filldedent('''
```
`TypeError` would make more sense here.

smichr replied:
```suggestion
            raise TypeError(filldedent('''
```

smichr commented on sympy/core/basic.py:
```diff
@@ -1258,6 +1258,33 @@ def has(self, *patterns):
+        False
+        """
+        # protect O(1) containment check by requiring:
+        if not type(s) in (dict, set):
+            raise ValueError('expecting set or dict argument')
```
```suggestion
        if type(s) is not set:
            raise TypeError('expecting set argument')
```

smichr commented on sympy/core/tests/test_expr.py:
```diff
@@ -1624,12 +1625,31 @@ def test_has_free():
+    assert Piecewise((s, x > 3), (4, True)).has_free(s)
+    assert not Piecewise((1, x > 3), (4, True)).has_free(s)
+    # can't make set of these, but fallback will handle
+    assert not x.has_free(y, [])
```
```suggestion
    assert not x.has_free(y, set())
```

smichr commented on sympy/core/tests/test_expr.py:
```diff
@@ -1624,12 +1625,31 @@ def test_has_free():
+    assert (x + 1 + y).has_free(x + 1)
+    assert not (x + 2 + y).has_free(x + 1)
+    assert (2 + 3*x*y).has_free(3*x)
+    raises(ValueError, lambda: x.has_free({x, y}))
```
```suggestion
    raises(TypeError, lambda: x.has_free({x, y}))
```

smichr commented on sympy/core/tests/test_expr.py:
```diff
@@ -1624,12 +1625,31 @@ def test_has_free():
+    assert Piecewise((s, x > 3), (4, True)).has_free(s)
+    assert not Piecewise((1, x > 3), (4, True)).has_free(s)
+    # can't make set of these, but fallback will handle
+    assert not x.has_free(y, [])
```
```suggestion
    raises(TypeError, lambda: x.has_free(y, []))
```

smichr commented on sympy/core/basic.py:
```diff
@@ -1284,8 +1311,27 @@ def has_free(self, *patterns):
+                non-Basic iterable. Don't forget to unpack
+                iterables: `eq.has_free(*patterns)`'''))
+        # try quick test first
+        try:
+            s = set(patterns)
+        except TypeError:
+            pass  # patterns had a non-hashable element
```
```suggestion
        s = set(patterns)
```

smichr commented on sympy/core/basic.py:
```diff
@@ -1284,8 +1311,24 @@ def has_free(self, *patterns):
+                iterables: `eq.has_free(*patterns)`'''))
+        # try quick test first
+        s = set(patterns)
+        else:
+            rv = self.has_xfree(s)
+            if rv:
+                return rv
```
```suggestion
        rv = self.has_xfree(s)
        if rv:
            return rv
```

smichr commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -2785,8 +2797,11 @@ def test_linear_coeffs():
     assert linear_coeffs(a*(x + y), x, y) == [a, a, 0]
     assert linear_coeffs(1.0, x, y) == [0, 0, 1.0]
+    # don't include coefficients of 0
+    assert linear_coeffs(Eq(x, x + y), x, y, dict=True) == {y: 1}
```
```suggestion
    assert linear_coeffs(Eq(x, x + y), x, y, dict=True) == {y: -1}
```

smichr commented on sympy/core/tests/test_expr.py:
```diff
@@ -1624,12 +1625,31 @@ def test_has_free():
+    assert (x + 1).has_xfree({x})
+    assert ((x + 1)**2).has_xfree({x + 1})
+    assert not (x + y + 1).has_xfree({x + 1})
+    raises(ValueError, lambda: x.has_xfree(x))
+    raises(ValueError, lambda: x.has_xfree([x]))
```
```suggestion
    raises(TypeError, lambda: x.has_xfree(x))
    raises(TypeError, lambda: x.has_xfree([x]))
```

smichr committed:
code review changes

smichr commented:
Any final thoughts, @oscarbenjamin ?

oscarbenjamin commented:
I think this looks good. Let's get it in!
